### DISCUSSION TITLE
crnn_vgg16_bn model - Link for downloading

### DISCUSSION THREAD

**Discussion by gganes3** [mindee/doctr#676] [Q&A] [answered]
Hi All, 

Just like the dbresnet (https://github.com/mindee/doctr/releases/download/v0.2.0/db_resnet50-adcafc63.zip) model file, I am trying to download the crnn_vgg16_bn model to use it offline. Could you please provide me the link?

--
Regards, 
Gokul.

**Comment by charlesmindee** [comment]
Hi @gganes3,
the download link is in the cfg of the model: 
[here for the pytorch version](https://github.com/mindee/doctr/blob/main/doctr/models/recognition/crnn/pytorch.py) and
[here for the tensorflow version](https://github.com/mindee/doctr/blob/main/doctr/models/recognition/crnn/tensorflow.py)

Have a nice day!

**Comment by fg-mindee** [comment, marked as answer]
Hello @gganes3 :wave: 

I'm not sure where you found the URL of `db_resnet50` but it's exactly the same process for other models!
To add up on Charles' answer, I'd recommend the following if you happen to be looking for other model's checkpoint URLs:
- navigate to the "models" page of the documentation: https://mindee.github.io/doctr/latest/models.html
- find the model you're interested in and click on "SOURCE" which will take you to the source code. 
- at the top of the page you'll see the URLs for all pretrained models :+1: 

If you're familiar with github, no need to go through the documentation, and you will find the crnn URLs:
- for pytorch here: https://github.com/mindee/doctr/blob/main/doctr/models/recognition/crnn/pytorch.py#L29
- for TensorFlow there: https://github.com/mindee/doctr/blob/main/doctr/models/recognition/crnn/tensorflow.py#L28

Take care!

**Reply by gganes3** [reply]
Thanks for the quick response. One more question, I am trying to load these models and use it without calling ocr_predictor because I want to use it in a place where github links can't be accessed. I am using code present at this link. https://github.com/mindee/doctr/discussions/568
But I am getting the error.  Can you please help here as well? Thanks in advance.
**ModuleNotFoundError: No module named 'doctr.models.predictor'**

**Reply by fg-mindee** [reply]
Mmmh that is strange, I'm guessing you don't have a recent docTR version!
Could you run this and post the output here please?
```shell
import doctr
print(doctr.__version__)
```

I'd suggest updating to the latest version if that's the case:
```shell
pip install python-doctr --upgrade
```

**Reply by gganes3** [reply]
After updating the doctr version and below code results in `UnpicklingError: invalid load key, 'm'`. 
Can you please help here?
```python
det_params = torch.load(".cache/doctr/models/db_resnet50-adcafc63/checkpoint", map_location="cpu")
reco_params = torch.load(".cache/doctr/models/crnn_vgg16_bn-748c855f/checkpoint", map_location="cpu")
det_model = torch.load("cache/doctr/models/db_resnet50-adcafc63/weights.data-00000-of-00001",map_location="cpu")
reco_model = torch.load("cache/doctr/models/crnn_vgg16_bn-748c855f/weights.data-00000-of-00001",map_location="cpu")
det_model.load_state_dict(det_params)
reco_model.load_state_dict(reco_params)
det_predictor = DetectionPredictor(PreProcessor((1024, 1024), batch_size=1, mean=(0.798, 0.785, 0.772), std=(0.264, 0.2749, 0.287)), det_model)
reco_predictor = RecognitionPredictor(PreProcessor((32, 128), preserve_aspect_ratio=True, batch_size=32, mean=(0.694, 0.695, 0.693), std=(0.299, 0.296, 0.301)), reco_model)

model = OCRPredictor(det_predictor, reco_predictor)
```

**Reply by fg-mindee** [reply]
I can take a look! Could you share the full traceback of the error please? :pray: 
(the entire cascade message leading to the error, that shows which submodules the error was originally from)

**Reply by gganes3** [reply]
```
      3         # Load model
----> 4         det_params = torch.load(".cache/doctr/models/db_resnet50-adcafc63/checkpoint", map_location="cpu")
      5         reco_params = torch.load(".cache/doctr/models/crnn_vgg16_bn-748c855f/checkpoint", map_location="cpu")
      6         det_model = torch.load(".cache/doctr/models/db_resnet50-adcafc63/weights.data-00000-of-00001",map_location="cpu")

/home/cdsw/.local/lib/python3.6/site-packages/torch/serialization.py in load(f, map_location, pickle_module, **pickle_load_args)
    606                     return torch.jit.load(opened_file)
    607                 return _load(opened_zipfile, map_location, pickle_module, **pickle_load_args)
--> 608         return _legacy_load(opened_file, map_location, pickle_module, **pickle_load_args)
    609 
    610 

/home/cdsw/.local/lib/python3.6/site-packages/torch/serialization.py in _legacy_load(f, map_location, pickle_module, **pickle_load_args)
    775             "functionality.")
    776 
--> 777     magic_number = pickle_module.load(f, **pickle_load_args)
    778     if magic_number != MAGIC_NUMBER:
    779         raise RuntimeError("Invalid magic number; corrupt file?")

UnpicklingError: invalid load key, 'm'.
```

Here is the full traceback. I am not sure about this pickle.load and am pretty new to it. Thanks a lot for helping. :)

**Reply by fg-mindee** [reply]
This might be out of my skills. But I just realized something: are you positive that this is a PyTorch checkpoint you're loading here?
Considering the "invalid load key", and that you haven't yet tried to load it on the model, I can only see the explanation that the checkpoint is not correctly formatted for PyTorch. So my guess is that it's either corrupted or that it's TF checkpoint (or something else) 

How did you save your checkpoint by the way? the default method from the reference scripts?

**Reply by gganes3** [reply]
Sorry, it was my mistake.  Thanks for giving me the right direction. I was not using the torch model. I downloaded the .pt files and now I can load the .pt files. However, one more question, how should I instantiate the det_model and rec_model in the below code. 

       # Instantiate your model
        **det_model= ??
        reco_model = ??**
        det_params = torch.load(r"C:\Projects\db_resnet50-ac60cadc.pt", map_location="cpu")
        reco_params = torch.load(r"C:\Projects\crnn_vgg16_bn-9762b0b0.pt", map_location="cpu")
        # det_model.load_state_dict(det_params)
        # reco_model.load_state_dict(reco_params)

        # Ask the preprocessor of each task to resize and normalize similarly to your training
        # cf. https://github.com/mindee/doctr/blob/main/references/detection/train_pytorch.py#L94 & https://github.com/mindee/doctr/blob/main/references/detection/train_pytorch.py#L109
        det_predictor = DetectionPredictor(
            PreProcessor((1024, 1024), batch_size=1, mean=(0.798, 0.785, 0.772), std=(0.264, 0.2749, 0.287)), det_model)
        # cf. https://github.com/mindee/doctr/blob/main/references/recognition/train_pytorch.py#L97 & https://github.com/mindee/doctr/blob/main/references/recognition/train_pytorch.py#L111
        reco_predictor = RecognitionPredictor(
            PreProcessor((32, 128), preserve_aspect_ratio=True, batch_size=32, mean=(0.694, 0.695, 0.693),
                         std=(0.299, 0.296, 0.301)), reco_model)

        model = OCRPredictor(det_predictor, reco_predictor)
        result = model([image_rgb])

**Reply by fg-mindee** [reply]
Rather easy then :) 

```python
from doctr.models.detection import db_resnet50
from doctr.models.recognition import crnn_vgg16_bn

det_model = db_resnet50(pretrained=False, pretrained_backbone=False)
reco_model = crnn_vgg16_bn(pretrained=False, pretrained_backbone=False)
```

and assuming you haven't changed the vocab, or any other params of both models

**Reply by gganes3** [reply]
Thanks, but I get this error _**'det_model.load_state_dict([det_params])
AttributeError: 'DBNet' object has no attribute 'load_state_dict''**_

      # Instantiate your model
        det_model = db_resnet50(pretrained=False)
        reco_model = crnn_vgg16_bn(pretrained=False)
        det_params = torch.load(r"C:\Projects\db_resnet50-ac60cadc.pt", map_location="cpu")
        reco_params = torch.load(r"C:\Projects\crnn_vgg16_bn-9762b0b0.pt", map_location="cpu")
        # det_model.load_state_dict()
        det_model.load_state_dict(det_params)
        reco_model.load_state_dict(reco_params)


        det_predictor = DetectionPredictor(
            PreProcessor((1024, 1024), batch_size=1, mean=(0.798, 0.785, 0.772), std=(0.264, 0.2749, 0.287)), det_model)

        reco_predictor = RecognitionPredictor(
            PreProcessor((32, 128), preserve_aspect_ratio=True, batch_size=32, mean=(0.694, 0.695, 0.693),
                         std=(0.299, 0.296, 0.301)), reco_model)

        model = OCRPredictor(det_predictor, reco_predictor)

**Reply by fg-mindee** [reply]
I'm not able to reproduce this locally, so my guess is:
- either you don't have the correct version of docTR installed (in this case, update it)
- or you're not running docTR with the correct backend (by default it's TensorFlow). To enable this, add this env variable `USE_TORCH` at the very top of your script:
```python

import os

os.environ['USE_TORCH'] = '1'
```

Hopefully that will do it :+1:
